libretro/LibretroGraphicsContext.h

```diff
@@ -0,0 +1,139 @@
+#pragma once
+#include <atomic>
+
+#include "libretro/libretro.h"
+#include "Common/GraphicsContext.h"
+#include "thin3d/thin3d_create.h"
+
+#include "Core/System.h"
+#include "GPU/GPUState.h"
+
+class LibretroGraphicsContext : public GraphicsContext {
+	public:
+	LibretroGraphicsContext() {}
+	~LibretroGraphicsContext() override {}
+
+	virtual bool Init() = 0;
+	virtual void SetRenderTarget() {}
+	virtual GPUCore GetGPUCore() = 0;
+	virtual const char *Ident() = 0;
+
+	void Shutdown() override
+	{
+		DestroyDrawContext();
+		PSP_CoreParameter().thin3d = nullptr;
+	}
+	void SwapInterval(int interval) override {}
+	void Resize() override {}
+
+	virtual void CreateDrawContext() {}
+	virtual void DestroyDrawContext()
+	{
+		if (!draw_)
+			return;
+		draw_->HandleEvent(Draw::Event::LOST_BACKBUFFER, -1, -1);
+		delete draw_;
+		draw_ = nullptr;
+	}
+	Draw::DrawContext *GetDrawContext() override { return draw_; }
+
+	static LibretroGraphicsContext *CreateGraphicsContext();
+
+	static retro_video_refresh_t video_cb;
+
+	protected:
+	Draw::DrawContext *draw_ = nullptr;
+};
+
+class LibretroHWRenderContext : public LibretroGraphicsContext {
+	public:
+	LibretroHWRenderContext(retro_hw_context_type context_type, unsigned version_major = 0, unsigned version_minor = 0);
+	bool Init() override;
+	void SetRenderTarget() override {}
+	void SwapBuffers() override
+	{
+		if (gstate_c.skipDrawReason)
+			video_cb(NULL, 0, 0, 0);
+		else
+			video_cb(RETRO_HW_FRAME_BUFFER_VALID, PSP_CoreParameter().pixelWidth, PSP_CoreParameter().pixelHeight, 0);
+	}
+	virtual void ContextReset();
+	virtual void ContextDestroy();
+
+	protected:
+	retro_hw_render_callback hw_render_ = {};
+};
+
+extern "C" retro_hw_get_proc_address_t libretro_get_proc_address;
+
+#ifdef _WIN32
+class LibretroD3D9Context : public LibretroHWRenderContext {
+	public:
+	LibretroD3D9Context() : LibretroHWRenderContext(RETRO_HW_CONTEXT_DIRECT3D, 9) {}
+	bool Init() override { return false; }
+#if 0
```

Comment:
Can we get rid of all these ```#if 0``` ?  Also they're mis-indented.

Reply:
they are just visual reminders about things that need to be added/fixed. I can get rid of them if that's preferred.
I'll fix the indent, I relied too much on clang-format heh.